staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -640,15 +666,70 @@ func (r *crdHandler) updateCustomResourceDefinition(oldObj, newObj interface{})
 		tearDownFinishedCh = r.removeStorageLocked(newCRD.UID)
 	}
 
-	// Update storage version with the latest info in the watch event
+	r.queueSVUpdate(newCRD, tearDownFinishedCh)
+}
+
+func (r *crdHandler) deleteCustomResourceDefinition(obj interface{}) {
+	r.removeDeadStorage()
+}
+
+func (r *crdHandler) queueSVUpdate(crd *apiextensionsv1.CustomResourceDefinition, tearDownFinishedCh <-chan struct{}) {
 	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
 		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
-		ctx := apirequest.NewContext()
-		err := r.storageVersionManager.UpdateStorageVersion(ctx, newCRD, tearDownFinishedCh, nil, nil)
-		if err != nil {
-			klog.Errorf("error updating storage version for %v: %v", newCRD, err)
-		}
+		r.recordLatestSVUpdateInfoWithTeardown(crd, tearDownFinishedCh)
+		r.storageVersionUpdaterQueue.Add(crd.UID)
+	}
+}
+
+func (r *crdHandler) runSVUpdateLoop(stopCh <-chan struct{}) {
+	go wait.Until(r.svUpdateWorker, time.Second, stopCh)
+	defer r.storageVersionUpdaterQueue.ShutDown()
+	<-stopCh
+}
+
+func (r *crdHandler) svUpdateWorker() {
+	// does this guarantee that all events are being processed as they come in?
+	for r.processLatestSVUpdate() {
+	}
+}
+
+func (r *crdHandler) processLatestSVUpdate() bool {
+	ctx := context.TODO()
+	key, quit := r.storageVersionUpdaterQueue.Get()
+	if quit {
+		return false
+	}
+	defer r.storageVersionUpdaterQueue.Done(key)
+
+	val, _ := r.latestStorageVersionUpdateInfo.Load(key)
+	latestSVUpdateInfo := val.(*storageVersionUpdateInfo)
+
+	// TODO: unsure if we need to pass ctx here.
+	err := r.storageVersionManager.UpdateStorageVersion(ctx, latestSVUpdateInfo.crd, latestSVUpdateInfo.updateChannels.teardownFinishedCh, latestSVUpdateInfo.updateChannels.processedCh, latestSVUpdateInfo.updateChannels.errCh)
+	if err == nil {
+		// TODO: fill me
+	}
+
+	return true
+}
+
+func (r *crdHandler) recordLatestSVUpdateInfoWithTeardown(crd *apiextensionsv1.CustomResourceDefinition, tearDownFinishedCh <-chan struct{}) {
+	svUpdateInfo := &storageVersionUpdateInfo{
+		crd:     crd,
+		timeout: time.Now().Add(storageVersionUpdateTimeout),
+		updateChannels: &storageVersionUpdateChannels{
+			processedCh:        make(chan struct{}),
+			errCh:              make(chan struct{}),
+			teardownFinishedCh: tearDownFinishedCh,
+		},
 	}
+
+	// overwrite existing SVUpdateInfo with latest update event.
+	r.updateLatestStorageVersionUpdateInfo(crd.UID, svUpdateInfo)
+}
+
+func (r *crdHandler) updateLatestStorageVersionUpdateInfo(crdUID types.UID, svUpdateInfo *storageVersionUpdateInfo) {
+	r.latestStorageVersionUpdateInfo.Store(crdUID, svUpdateInfo)
```

Comment:
IIUC we are only tracking the latest teardown. If there are multiple versions being torn down, do we block new CR writes until all the versions have been torn down?

Reply:
So, we will be blocking CR writes on only the latest SV update to finish meaning we will wait for the latest teardown to finish. 
If  all multiple storageversions updates are in progress, v1 -> v2 -> v3 and a CR request happens after v3 update was requested, CR would be blocked on v3 update to finish, v3 update would be blocked on v2 teardown to finish and v2 update will be blocked on v1 teardown to finish. So in a way we will be waiting for all versions to teardown.

If multiple rapid CRD updates come in, and the map gets updated from v1 -> v3 directly, the new CR would be blocked on just v1 teardown to finish.

Reply:
So I guess at any given time, because we only process the latest SV update, there can only be a single teardown in progress?